Update Nederlandse Spoorwegen integration documentation and improve c… (#39826)
home-assistant/home-assistant.github.io: source/_integrations/nederlandse_spoorwegen.markdown

ha_iot_class: Cloud Polling
ha_release: 0.57
ha_codeowners:
  - '@heindrichpaul'
  - '@YarmoM'
ha_domain: nederlandse_spoorwegen
ha_platforms:
  - sensor
ha_integration_type: integration

> ha_quality_scale: bronze

Reviewer: not yet

---

The **Nederlandse Spoorwegen (NS)** {% term integration %} provides real-time information about Dutch train schedules using the [NS API](https://apiportal.ns.nl/). This integration allows you to monitor departure times, delays, and travel information for your regular routes.

## Prerequisites

To use this integration, you need an API key from the NS API Portal.

### Obtaining an API Key

1. Create an account on the [NS API Portal](https://apiportal.ns.nl/).
2. Request an API key for the `Reisinformatie` API, which is part of the `Ns-App` product.

{% include integrations/config_flow.md %}

### Managing routes

After adding the integration, you can manage your travel routes:

1. Go to **Settings** > **Devices & Services**
2. Find the Nederlandse Spoorwegen integration
3. Click **Configure**

> 4. Add or remove routes as needed

Reviewer 2: ```suggestion
1. Go to {% my integrations title="**Settings** > **Devices & services**" %}.
2. Find the **Nederlandse Spoorwegen** integration.
3. Click **Configure**.
4. Add or remove routes as needed.
```

Using a specific time only updates the route sensor during a time window around the chosen time. Outside this window, the route sensor's state is `unknown`. The window is from half an hour before the chosen time until half an hour after the chosen time. In this way, you can have multiple routes with specific trains before hitting the API usage limits.

## Migration from YAML

{% important %}
YAML configuration for Nederlandse Spoorwegen is now deprecated and will be removed in a future release.
{% endimportant %}

The integration automatically migrates existing YAML configuration and creates repair notifications to guide users. If you have an existing YAML configuration, the integration will automatically import your routes when you add the integration through the UI.

To complete the migration:

1. Remove the `nederlandse_spoorwegen:` section from your `configuration.yaml`
2. Remove any `sensor:` entries with `platform: nederlandse_spoorwegen`

> 3. Restart Home Assistant to clear the repair notifications

Reviewer 2: ```suggestion
1. Remove the `nederlandse_spoorwegen:` section from your `configuration.yaml`.
2. Remove any `sensor:` entries with `platform: nederlandse_spoorwegen`.
3. Restart Home Assistant to clear the repair notifications.
```

> Your existing routes and settings are automatically preserved in the new UI-based configuration.

Reviewer: We don't document deprecated config

## Migration from YAML

{% important %}
YAML configuration for Nederlandse Spoorwegen is now deprecated and will be removed in a future release.
{% endimportant %}

The integration automatically migrates existing YAML configuration and creates repair notifications to guide users. If you have an existing YAML configuration, the integration will automatically import your routes when you add the integration through the UI.

To complete the migration:

1. Remove the `nederlandse_spoorwegen:` section from your `configuration.yaml`
2. Remove any `sensor:` entries with `platform: nederlandse_spoorwegen`
3. Restart Home Assistant to clear the repair notifications

Your existing routes and settings are automatically preserved in the new UI-based configuration.

## Data source

The data is provided by Nederlandse Spoorwegen through their official API, ensuring high-quality and up-to-date information about train schedules, delays, and service disruptions.

## Troubleshooting

### Authentication errors

If you encounter authentication errors:

- Verify your API key is correct
- Ensure your NS API subscription is active

> - Check that you're using the correct API (Reisinformatie API)

Reviewer 2: ```suggestion
- Verify your API key is correct.
- Ensure your NS API subscription is active.
- Check that you're using the correct API (Reisinformatie API).
```

### Removing the integration

To completely remove the integration:

1. Go to **Settings** > **Devices & Services**
2. Find the Nederlandse Spoorwegen integration
3. Click the three-dot menu and select **Delete**

> 4. Confirm the removal

Reviewer 2: ```suggestion
This integration follows standard integration removal.

{% include integrations/remove_device_service.md %}
```
There is a standard [text block for integration removal](https://github.com/home-assistant/home-assistant.io/blob/current/source/_integrations/_integration_docs_template.markdown#removing-the-integration)

ha_iot_class: Cloud Polling
ha_release: 0.57
ha_codeowners:
  - '@heindrichpaul'
  - '@YarmoM'
ha_domain: nederlandse_spoorwegen
ha_platforms:
  - sensor
ha_integration_type: integration
ha_quality_scale: bronze
---

The **Nederlandse Spoorwegen (NS)** {% term integration %} provides real-time information about Dutch train schedules using the [NS API](https://apiportal.ns.nl/). This integration allows you to monitor departure times, delays, and travel information for your regular routes.

> {% include integrations/config_flow.md %}

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
## Prerequisites

```

To use this integration, you need an API key from the NS API Portal.

### Obtaining an API Key

> To use this integration, you need an API key from NS:

Reviewer 2: ```suggestion
```

1. Create an account on the [NS API Portal](https://apiportal.ns.nl/).
2. Request an API key for the `Reisinformatie` API, which is part of the `Ns-App` product.

> ## Configuration
> 
> ### Adding the integration
> 
> 1. In the Home Assistant UI, go to **Settings** > **Devices & Services**
> 2. Click **Add Integration**
> 3. Search for and select **Nederlandse Spoorwegen (NS)**
> 4. Enter your NS API key
> 5. Click **Submit**

Reviewer 2: ```suggestion
{% include integrations/config_flow.md %}
```

ha_release: 0.57
ha_codeowners:
  - '@YarmoM'
  - '@heindrichpaul'
ha_domain: nederlandse_spoorwegen
ha_platforms:
  - sensor
ha_integration_type: integration
related:
  - docs: /docs/configuration/
    title: Configuration file
ha_quality_scale: silver
---

> This {% term integration %} provides real-time timetable information for [Nederlandse Spoorwegen](https://www.ns.nl/) (NS), the Dutch national railway, directly in Home Assistant.

Reviewer 2: ```suggestion
The **Nederlandse Spoorwegen (NS)** {% term integration %} provides real-time timetable information for [Nederlandse Spoorwegen](https://www.ns.nl/) (NS), the Dutch national railway, directly in Home Assistant.
```

## Getting Started

### Obtain an API Key

To use this integration, you need an API key from NS:

1. Create an account on the [NS API Portal](https://apiportal.ns.nl/).
2. Request an API key for the `Reisinformatie` API, which is part of the `Ns-App` product.

### Configuration Methods

You can set up the NS integration using either the Home Assistant UI or by editing your `configuration.yaml` file.

#### UI Configuration Flow (Recommended)

1. Go to **Settings > Devices & Services** in Home Assistant.
2. Click **Add Integration** and search for "Nederlandse Spoorwegen".
3. Enter your NS API key when prompted.
4. Add one or more train routes by specifying the departure station, arrival station, and optionally a via station.
5. Save your configuration. The integration will create sensors for each route.

You can edit, add, or remove routes at any time using the integration's options flow in the UI. If your API key changes or expires, use the reauthentication flow to update it without removing your integration.

#### YAML Configuration (Advanced)

```yaml
# Example configuration.yaml entry
sensor:

>   - platform: nederlandse_spoorwegen

Reviewer 2: is there anything I can do in the YAML config that I can't do from the UI?

Author: No. So technically we can remove the mention of the configuration completely.

Reviewer 2: ah, ok. yes, if there is no additional functionality in YAML, I think we don't usually document anymore that you can go to the config file. The information could move into a [configuration_basic block](https://developers.home-assistant.io/docs/documenting/integration-docs-examples#configuration_basic-block)

Author: I have changed it to use the basic configuration

related:
  - docs: /docs/configuration/
    title: Configuration file
ha_quality_scale: silver
---

The **Nederlandse Spoorwegen (NS)** {% term integration %} provides real-time timetable information for [Nederlandse Spoorwegen](https://www.ns.nl/) (NS), the Dutch national railway, directly in Home Assistant.

## Prerequisites

### Obtaining an API Key

To use this integration, you need an API key from NS:

1. Create an account on the [NS API Portal](https://apiportal.ns.nl/).
2. Request an API key for the `Reisinformatie` API, which is part of the `Ns-App` product.

Station codes are required. You can look them up in the [list of railway stations in the Netherlands](https://nl.wikipedia.org/wiki/Lijst_van_spoorwegstations_in_Nederland).

> {% include integrations/config_flow.md %}

Reviewer 2: ```suggestion
### Finding station codes

Station codes are required. You can look them up in the [list of railway stations in the Netherlands](https://nl.wikipedia.org/wiki/Lijst_van_spoorwegstations_in_Nederland).

{% include integrations/config_flow.md %}

```

related:
  - docs: /docs/configuration/
    title: Configuration file
ha_quality_scale: silver
---

The **Nederlandse Spoorwegen (NS)** {% term integration %} provides real-time timetable information for [Nederlandse Spoorwegen](https://www.ns.nl/) (NS), the Dutch national railway, directly in Home Assistant.

## Prerequisites

> 

Reviewer 2: ```suggestion

To use this integration, you need an API key and station codes.

```

### Obtaining an API Key

To use this integration, you need an API key from NS:

1. Create an account on the [NS API Portal](https://apiportal.ns.nl/).
2. Request an API key for the `Reisinformatie` API, which is part of the `Ns-App` product.

> ### Finding station codes

Reviewer 2: ```suggestion

### Finding station codes
```

related:
  - docs: /docs/configuration/

>     title: Configuration file

Reviewer 2: ```suggestion
```

ha_quality_scale: silver
---

> 

Reviewer 2: @heindrichpaul, I restructured a little bit to align with the structure of other integrations. I also added the removal section. Could you double-check if it is still all correct? 
And regarding the removal, are there any other steps needed? Should I delete the account on NS, for example?

Author: @c0ffeeca7 I took a look and it look good.

related:
  - docs: /docs/configuration/
    title: Configuration file
ha_quality_scale: silver
---

The **Nederlandse Spoorwegen (NS)** {% term integration %} provides real-time timetable information for [Nederlandse Spoorwegen](https://www.ns.nl/) (NS), the Dutch national railway, directly in Home Assistant.

> ## Getting Started

Reviewer 2: ```suggestion
## Prerequisites
```

> ### Obtain an API Key

Reviewer 2: ```suggestion
### Obtaining an API Key
```

To use this integration, you need an API key from NS:

1. Create an account on the [NS API Portal](https://apiportal.ns.nl/).
2. Request an API key for the `Reisinformatie` API, which is part of the `Ns-App` product.

> ### Configuration
> 
> You can set up the NS integration using the Home Assistant UI.

Reviewer 2: ```suggestion
Station codes are required. You can look them up in the [list of railway stations in the Netherlands](https://nl.wikipedia.org/wiki/Lijst_van_spoorwegstations_in_Nederland).

{% include integrations/config_flow.md %}
```

> 
> #### UI Configuration Flow
> 
> 1. Go to **Settings > Devices & Services** in Home Assistant.
> 2. Click **Add Integration** and search for "Nederlandse Spoorwegen".

Reviewer 2: ```suggestion
```

3. Enter your NS API key when prompted.
4. Add one or more train routes by specifying the departure station, arrival station, and optionally a via station.

> 5. Save your configuration. The integration will create sensors for each route.

Reviewer 2: ```suggestion
1. Enter your NS API key when prompted.
2. Add one or more train routes by specifying the departure station, arrival station, and optionally a via station.
3. Save your configuration. The integration will create sensors for each route.
```

You can edit, add, or remove routes at any time using the integration's options flow in the UI. If your API key changes or expires, use the reauthentication flow to update it. For other configuration changes, use the reconfiguration flow, both without removing your integration.

> {% include integrations/config_flow.md %}

Reviewer 2: ```suggestion
```

{% configuration_basic %}
api_key:
    description: "API key obtained from the NS API Portal."
from:
    description: "Departure station code (for example `Rtd`)."
to:
    description: "Arrival station code (for example `Asd`)."
via:
    description: "Optional intermediate station code."
name:
    description: "A descriptive name for this route."
{% endconfiguration_basic %}

> ### Station codes

Reviewer 2: ```suggestion
```

> Station codes are required. You can look them up in the [list of railway stations in the Netherlands](https://nl.wikipedia.org/wiki/Lijst_van_spoorwegstations_in_Nederland).

Reviewer 2: ```suggestion
```

The window is from half an hour before the chosen time until half an hour after the chosen time.
In this way, you can have multiple routes with specific trains before hitting the FUP threshold for using NS API.

### Data source

Data is provided by [Nederlandse Spoorwegen](https://www.ns.nl/).

> 

Reviewer 2: ```suggestion
```

The window is from half an hour before the chosen time until half an hour after the chosen time.
In this way, you can have multiple routes with specific trains before hitting the FUP threshold for using NS API.

### Data source

Data is provided by [Nederlandse Spoorwegen](https://www.ns.nl/).

> This integration meets the [Silver quality-scale requirements](https://developers.home-assistant.io/docs/integration_quality_scale_index/), ensuring reliable configuration via the Home Assistant UI and robust error handling.

Reviewer 2: ```suggestion
```
No need to mention this here. The quality scale is part of the metadata and will be shown in the Info block.
![image](https://github.com/user-attachments/assets/52f41f61-06bc-4df6-a7d5-7cf41dcbc7c2)

The window is from half an hour before the chosen time until half an hour after the chosen time.
In this way, you can have multiple routes with specific trains before hitting the FUP threshold for using NS API.

### Data source

Data is provided by [Nederlandse Spoorwegen](https://www.ns.nl/).

> This integration meets the [Silver quality-scale requirements](https://developers.home-assistant.io/docs/integration_quality_scale_index/), ensuring reliable configuration via the Home Assistant UI and robust error handling.

Reviewer 2: ```suggestion
This integration meets the [Silver quality-scale requirements](https://developers.home-assistant.io/docs/integration_quality_scale_index/), ensuring reliable configuration via the Home Assistant UI and robust error handling.

## Removing the integration

This integration follows standard integration removal.

{% include integrations/remove_device_service.md %}
```

      required: false
      type: string
    time:
      description: Optional time to search for a specific train (format `hh:mm:ss`, for example `08:06:00`).
      required: false
      type: time
{% endconfiguration %}

{% note %}
After adding or updating your configuration, restart Home Assistant to apply the changes.
{% endnote %}

### Station codes

> 

Reviewer 2: the station codes, ideally, do I know them before installing the integration? Would it make sense to mention that in the Prerequisites section?

Author: No it is a foot note in the documentation. You also have a link in the config flow if you need it at the time of configuration
